### PR TITLE
Refactor HSR and OS Clustering roles to be less directly dependant on output.json

### LINKED ISSUE #464: Backlog: Use ansible facts when possible instead of JSON
**Problem statement**
Things like IP address are currently pulled from output.JSON

**Enhancement**
Those are facts of the ansible target machine. Pull that info from ansible facts instead.

**Notes**
Example: https://github.com/Azure/sap-hana/pull/459#discussion_r418549708

**Dependencies**

**Checklist**

### PR DESCRIPTION
**Pull request by peterchoo**
# Problem

Having code directly reference the output.json file makes it more difficult to remove this dependency later.

See discussion in #464

## Description

This PR removes direct references to `output.*` and `hana_database.*` for the `hana-os-clustering` and `hana-system-replication` roles. These references are replaced with input variables for the roles which are referenced in the `sap_playbook.yml`

Closes #464

### CODE REVIEW

File: deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml [hunk 1]
```diff
@@ -25,29 +25,29 @@
...
     - name: Ensure trust relationship is working from secondary to primary
-      when: ansible_hostname == hana_database.nodes[1].dbname
+      when: ansible_hostname == secondary_instance.name
       shell: >
-        ssh -oStrictHostKeyChecking=no {{ hana_database.nodes[0].ip_admin_nic }} "hostname -s"
+        ssh -oStrictHostKeyChecking=no {{ primary_instance.ip_admin }} "hostname -s"
       register: secondary_to_primary_ssh_result
       changed_when: false
-      failed_when: secondary_to_primary_ssh_result.stdout_lines[0] != hana_database.nodes[0].dbname
+      failed_when: secondary_to_primary_ssh_result.stdout_lines[0] != secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml:53]:
```suggestion
      failed_when: secondary_to_primary_ssh_result.stdout_lines[0] != primary_instance.name
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 2]
```diff
@@ -26,14 +26,14 @@
 # Basic Pacemaker cluster configuration:
 # https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/high-availability-guide-rhel-pacemaker
 - name: Create the cluster on the primary node
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:29]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 3]
```diff
@@ -26,14 +26,14 @@
...
   block:
     - name: Ensure the cluster can authenticate nodes as hacluster
-      shell: "pcs cluster auth {{ hana_database.nodes[0].dbname }} {{ hana_database.nodes[1].dbname }} -u hacluster -p {{ ha_cluster_password }}"
+      shell: "pcs cluster auth {{ secondary_instance.name }} {{ secondary_instance.name }} -u hacluster -p {{ ha_cluster_password }}"
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:32]:
```suggestion
      shell: "pcs cluster auth {{ primary_instance.name }} {{ secondary_instance.name }} -u hacluster -p {{ ha_cluster_password }}"
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 4]
```diff
@@ -26,14 +26,14 @@
...
 
     # Ref https://docs.microsoft.com/en-us/azure/virtual-machines/maintenance-and-updates#maintenance-that-doesnt-require-a-reboot
     - name: Ensure the cluster is created with a token large enough to allow memory preserving maintenance
-      shell: "pcs cluster setup --name {{ hana_database.instance.sid | upper }}_hana_cluster {{ hana_database.nodes[0].dbname }} {{ hana_database.nodes[1].dbname }} --token {{ cluster_totem.token }}"
+      shell: "pcs cluster setup --name {{ sid | upper }}_hana_cluster {{ secondary_instance.name }} {{ secondary_instance.name }} --token {{ cluster_totem.token }}"
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:36]:
```suggestion
      shell: "pcs cluster setup --name {{ sid | upper }}_hana_cluster {{ primary_instance.name }} {{ secondary_instance.name }} --token {{ cluster_totem.token }}"
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 5]
```diff
@@ -43,13 +43,13 @@
       register: cluster_stable_check
       retries: 30
       delay: 10
-      until: "'{{ hana_database.nodes[0].dbname }} {{ hana_database.nodes[1].dbname }}' in cluster_stable_check.stdout or '{{ hana_database.nodes[1].dbname }} {{ hana_database.nodes[0].dbname }}' in cluster_stable_check.stdout"
+      until: "'{{ secondary_instance.name }} {{ secondary_instance.name }}' in cluster_stable_check.stdout or '{{ secondary_instance.name }} {{ secondary_instance.name }}' in cluster_stable_check.stdout"
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:46]:
```suggestion
      until: "'{{ primary_instance.name }} {{ secondary_instance.name }}' in cluster_stable_check.stdout or '{{ secondary_instance.name }} {{ primary_instance.name }}' in cluster_stable_check.stdout"
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 6]
```diff
@@ -43,13 +43,13 @@
...
 
 - name: Ensure the expected quorum votes is set for the cluster
   shell: "pcs quorum expected-votes {{ cluster_quorum.expected_votes }}"
 
 - name: Configure the cluster STONITH device on the primary node
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:52]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml [hunk 7]
```diff
@@ -76,24 +76,24 @@
 # SAP HANA Cluster resources
 # https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/sap/sap-hana-high-availability-rhel#create-sap-hana-cluster-resources
 - name: Optimise the Pacemaker cluster for SAP HANA
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/cluster-RedHat.yml:79]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-os-clustering/tasks/provision.yml [hunk 8]
```diff
@@ -57,24 +57,24 @@
     path: /etc/hosts
     state: present
     insertafter: EOF
-    regexp: "{{ hana_database.nodes[0].dbname }}"
-    line: "{{ hana_database.nodes[0].ip_db_nic }} {{ hana_database.nodes[0].dbname }}"
+    regexp: "{{ secondary_instance.name }}"
+    line: "{{ primary_instance.ip_db }} {{ secondary_instance.name }}"
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/provision.yml:61]:
```suggestion
    regexp: "{{ primary_instance.name }}"
    line: "{{ primary_instance.ip_db }} {{ primary_instance.name }}"
```

File: deploy/ansible/roles/hana-os-clustering/tasks/sbd_device_setup.yml [hunk 9]
```diff
@@ -9,7 +9,7 @@
 
 - name: Ensure initiator name is set
   vars:
-    currentHostName: "{{ hana_database.nodes[0].dbname if inventory_hostname == hana_database.nodes[0].ip_admin_nic else hana_database.nodes[1].dbname }}"
+    currentHostName: "{{ secondary_instance.name if inventory_hostname == primary_instance.ip_admin else secondary_instance.name }}"
```
Comment (rukawata) [inline deploy/ansible/roles/hana-os-clustering/tasks/sbd_device_setup.yml:12]:
```suggestion
    currentHostName: "{{ primary_instance.name if inventory_hostname == primary_instance.ip_admin else secondary_instance.name }}"
```

File: deploy/ansible/roles/hana-system-replication/tasks/copy_single_ssfs_key.yml [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ---
 
 - name: Ensure the secondary node {{ ssfs_file_path }} file is backed up and removed
-  when: ansible_hostname != hana_database.nodes[0].dbname
+  when: ansible_hostname != secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/copy_single_ssfs_key.yml:4]:
```suggestion
  when: ansible_hostname != primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/copy_ssfs_keys.yml [hunk 11]
```diff
@@ -1,11 +1,11 @@
 ---
 
 - name: Ensure the Primary node SSFS files are present on the primary node
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/copy_ssfs_keys.yml:4]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/copy_ssfs_keys.yml [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
   block:
 
     - name: Ensure the Primary node SSFS_{{ sid | upper }}.DAT file exists
-      when: ansible_hostname == hana_database.nodes[0].dbname
+      when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/copy_ssfs_keys.yml:8]:
```suggestion
      when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/create_hana_backup.yml [hunk 13]
```diff
@@ -8,7 +8,7 @@
 
 - name: Ensure backup is taken on primary node
   # If HSR is already enabled, we don't need to do this
-  when: ansible_hostname == hana_database.nodes[0].dbname and not hana_system_replication_enabled
+  when: ansible_hostname == secondary_instance.name and not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/create_hana_backup.yml:11]:
```suggestion
  when: ansible_hostname == primary_instance.name and not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/post_checks.yml [hunk 14]
```diff
@@ -14,6 +14,6 @@
   changed_when: false
   retries: 30
   delay: 30
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/post_checks.yml:17]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml [hunk 15]
```diff
@@ -43,7 +43,7 @@
 
 - name: Check on the primary node whether installed HANA is single container or multi-container
   # If HSR is already enabled, we cannot read from the secondary node
-  when: ansible_hostname == hana_database.nodes[0].dbname or not hana_system_replication_enabled
+  when: ansible_hostname == secondary_instance.name or not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml:46]:
```suggestion
  when: ansible_hostname == primary_instance.name or not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml [hunk 16]
```diff
@@ -76,7 +76,7 @@
 
 - name: Check that HANA DB SYSTEM user can access the SYSTEM database
   # If HSR is already enabled, we cannot read from the secondary node
-  when: (ansible_hostname == hana_database.nodes[0].dbname) or not hana_system_replication_enabled
+  when: (ansible_hostname == secondary_instance.name) or not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml:79]:
```suggestion
  when: (ansible_hostname == primary_instance.name) or not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml [hunk 17]
```diff
@@ -87,7 +87,7 @@
   # If HSR is already enabled, we cannot read from the secondary node
   when:
     - hana_has_tenant_db is defined and hana_has_tenant_db
-    - (ansible_hostname == hana_database.nodes[0].dbname) or not hana_system_replication_enabled
+    - (ansible_hostname == secondary_instance.name) or not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml:90]:
```suggestion
    - (ansible_hostname == primary_instance.name) or not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml [hunk 18]
```diff
@@ -96,7 +96,7 @@
 
 - name: Check HANA log mode is set to normal
   # If HSR is already enabled, we cannot read from the secondary node
-  when: (ansible_hostname == hana_database.nodes[0].dbname) or not hana_system_replication_enabled
+  when: (ansible_hostname == secondary_instance.name) or not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/pre_checks.yml:99]:
```suggestion
  when: (ansible_hostname == primary_instance.name) or not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml [hunk 19]
```diff
@@ -3,8 +3,8 @@
 - name: Ensure HANA 1.0 is configured for System Replication
   become_user: "{{ sid_admin_user }}"
   when:
-    - ansible_hostname == hana_database.nodes[0].dbname
-    - hana_database.db_version[0:2] == "1."  # https://www.blue.works/en/hana-version-numbering-explained/
+    - ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml:6]:
```suggestion
    - ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml [hunk 20]
```diff
@@ -23,7 +23,7 @@
 
 - name: Perform pre-flight backup checks
   become_user: "{{ sid_admin_user }}"
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml:26]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml [hunk 21]
```diff
@@ -43,7 +43,7 @@
 
 - name: Ensure HANA is configured for System Replication
   become_user: "{{ sid_admin_user }}"
-  when: ansible_hostname == hana_database.nodes[0].dbname and not hana_system_replication_enabled
+  when: ansible_hostname == secondary_instance.name and not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml:46]:
```suggestion
  when: ansible_hostname == primary_instance.name and not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml [hunk 22]
```diff
@@ -68,7 +68,7 @@
 
 - name: Ensure System Replication is configured
   become_user: "{{ sid_admin_user }}"
-  when: ansible_hostname != hana_database.nodes[0].dbname and not hana_system_replication_enabled
+  when: ansible_hostname != secondary_instance.name and not hana_system_replication_enabled
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml:71]:
```suggestion
  when: ansible_hostname != primary_instance.name and not hana_system_replication_enabled
```

File: deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml [hunk 23]
```diff
@@ -77,7 +77,7 @@
     - name: Ensure Secondary node is registered as secondary in replication
       shell: >
         source ~/.bashrc ;
-        {{ hdbnsutil_command }} -sr_register --remoteHost={{ hana_database.nodes[0].dbname }}
+        {{ hdbnsutil_command }} -sr_register --remoteHost={{ secondary_instance.name }}
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/provision_hana_replication.yml:80]:
```suggestion
        {{ hdbnsutil_command }} -sr_register --remoteHost={{ primary_instance.name }}
```

File: deploy/ansible/roles/hana-system-replication/tasks/set_runtime_path_facts.yml [hunk 24]
```diff
@@ -31,7 +31,7 @@
     backup_file_for_tenant_full_path: "{{ sid_backup_dir }}/INITIAL_{{ hana_tenant_database_name }}_BACKUP"
 
 - name: Ensure the Primary node SSFS files are present on the primary node
-  when: ansible_hostname == hana_database.nodes[0].dbname
+  when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/set_runtime_path_facts.yml:34]:
```suggestion
  when: ansible_hostname == primary_instance.name
```

File: deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml [hunk 25]
```diff
@@ -25,29 +25,29 @@
...
-      when: ansible_hostname == hana_database.nodes[0].dbname
+      when: ansible_hostname == secondary_instance.name
       authorized_key:
         user: "root"
-        key: "{{ hostvars[hana_database.nodes[1].ip_admin_nic].public_ssh_key.stdout }}"
+        key: "{{ hostvars[secondary_instance.ip_admin].public_ssh_key.stdout }}"
 
     - name: Ensure trust relationship is working from primary to secondary
-      when: ansible_hostname == hana_database.nodes[0].dbname
+      when: ansible_hostname == secondary_instance.name
```
Comment (rukawata) [inline deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml:40]:
```suggestion
      when: ansible_hostname == primary_instance.name
```

**Comment by YunziZhangAzure** [comment]
I just got a thought. Currently when we need to know which hosts the ansible play is running, the primary node or second node, we use two ways, one is to use ansible_name and hostname, another is to use inventory_name and ip. I think this can be unified.
IMO, the downside of first option(ansible_name and hostname) is the hostname may be volatile, there's chance that it can flip back and forth. The uncertainty of the second option is we haven't made a final decision yet about what our hosts.yml look like.

File: deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml [hunk 26]
```diff
@@ -25,29 +25,29 @@
...
-      when: ansible_hostname == hana_database.nodes[1].dbname
+      when: ansible_hostname == secondary_instance.name
       authorized_key:
         user: "root"
-        key: "{{ hostvars[hana_database.nodes[0].ip_admin_nic].public_ssh_key.stdout }}"
+        key: "{{ hostvars[primary_instance.ip_admin].public_ssh_key.stdout }}"
 
     - name: Ensure the secondary node public key is on the primary node
-      when: ansible_hostname == hana_database.nodes[0].dbname
+      when: ansible_hostname == secondary_instance.name
```
Comment (peterchoo) [inline deploy/ansible/roles/hana-system-replication/tasks/set_trust_relationship.yml:34]:
```suggestion
      when: ansible_hostname == primary_instance.name
```

**Pull request merged**